### ISSUE TITLE
Policy remediation will not work after policy is deployed via manage-azure-policy action

### ISSUE THREAD

**Issue by fawohlsc** [Azure/manage-azure-policy#23]
**Steps to reproduce:**
1. Create Azure policy definition and assignment in Portal - policy effect has to support remediation, e.g. DeployIfNotExists
2. Export both into GitHub from Portal
3. Delete Azure policy definition and assignment in Portal, so that manage azure policy action has to setup everything from scratch
4. Run manage azure policy action, which will create the policy definition and assignment from scratch
5. Remediate one of the policies will cause the following error message:

```
Error message you get during policy remediation: "The client '<MANAGED_IDENTITY_USED_FOR_REMEDIATION>' with object id '<MANAGED_IDENTITY_USED_FOR_REMEDIATION>' does not have authorization to perform action 'Microsoft.Resources/deployments/validate/action'
 ver scope '/subscriptions/<SUBSCRIPTION_ID>/resourcegroups/1808a0d8-3365-4f66-bcd3-9c64625521c1/providers/Microsoft.Resources/deployments/PolicyDeployment_16080442889105320296' or the scope is invalid.
If access was recently granted, please refresh your credentials."
```
_(Managed identity and subscription id are masked for privacy reasons)_

**Root cause analysis:**
When policies are assigned with remediation, a managed identity is required to actually execute that remediation (i.e. DeployIfNotExists, Modify, etc. ). Hence the respective roles have to be assigned to that managed identity. When you look at the underlying REST API requests performed by Azure Portal or Azure PowerShell (`New-AzPolicyAssignment ... -AssignIdentity -Debug`) during policy assignment, additional requests are sent to setup the role assignments of the managed identity. On first glance, I could not find that additional request being sent out in the source code of the manage-azure-policy action (well-written code btw. 🥇 ). Also, when manually assigning the required role to the managed identity, the policy remediation works:
```
$managedIdentity = (Get-AzPolicyAssignment | Where-Object { $_.Properties.DisplayName -eq "<POLICY_NAME>"  }).Identity
New-AzRoleAssignment -ObjectId $identity.principalId -RoleDefinitionName "Network Contributor" -Scope /subscriptions/<SUBSCRIPTION_ID>
```
_(Subscription id is masked for privacy reasons)_

Finally, I also sent a sample policy definition and assignment to @raiyanalam to ease reproduction.

**Comment by fawohlsc** [comment]
Also the completion of the policy assignment, including the optional role assignment for policies with remediation, should be await by the manage-azure-policy action. See examples in Azure PowerShell:

- [NewAzurePolicyAssignmentCmdlet](https://github.com/Azure/azure-powershell/blob/5278d3e565eb4ce3da626ba34bf63005697793d0/src/Resources/ResourceManager/Implementation/Policy/NewAzurePolicyAssignment.cs#L177)
- [LongRunningOperationHelper](https://github.com/Azure/azure-powershell/blob/1bcbe7b1f7a3323ac98f7754ba03eeb6b45e79f2/src/Resources/ResourceManager/Components/LongRunningOperationHelper.cs#L112)

**Comment by tauhid621** [comment]
@fawohlsc Thanks for the detailed explanation and RCA. I spent some time in understanding the issue. We missed the case of adding role assignments to the managed identity. I believe adding role assignment step after policy definition/assignment update will resolve this. I am not sure if this applies to policy initiatives as well.

**Comment by fawohlsc** [comment]
Hey @tauhid621! First of all thanks for the response. You are absolutely right about the additional call to add the roll assignment for the managed identity. For long-running operations, the Azure REST API returns an HTTP status code 202 (Accepted) and an URL for the status update to determine when the operation is completed is returned. This basically means that you have to perform [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting) and periodically call the URL for the status update to wait for the operations to complete. In Azure PowerShell, this is already handled for you (See: [LongRunningOperationHelper.cs](https://github.com/Azure/azure-powershell/blob/1bcbe7b1f7a3323ac98f7754ba03eeb6b45e79f2/src/Resources/ResourceManager/Components/LongRunningOperationHelper.cs#L112)). To cut a long story short: When your action finishes, it does not mean that the policy definition and assignments were successfully deployed - it can still fail since the operation is still running in the background. Same applies when for role assignments. Makes sense?

**Comment by fawohlsc** [comment]
In my [repository](https://github.com/fawohlsc/azure-policy-testing/blob/main/.github/workflows/test-policies.yml#L52) you can see how I worked around that issue.

**Comment by tauhid621** [comment]
Closing the issue as fix was rolled out.

**Comment by blueboxes** [comment]
I am still seeing this issue, even with the owner role specified in the policy it fails with this error message which running remediation .

**Comment by raiyanalam** [comment]
@blueboxes : Thanks for the comment and we're sorry you are facing this. Can you email me at moala [at] microsoft [dot] com with additional details or logs that you saw. This will help us debug the issue.

**Comment by tauhid621** [comment]
@blueboxes Were there any error during role assignment when the action ran?

**Comment by blueboxes** [comment]
@raiyanalam thanks I have sent an email.

I have written a custom policy to auto-apply locks to storage accounts and assigned it to a management group.

It was not working when I created new resources so I tried a remediation task on one resource in one subscription, this failed as unauthorised. I tried a number of time increasing the roles however even with owner it fails.

The only error message is in the remediation details.
